server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt

```diff
@@ -0,0 +1,151 @@
+import org.jetbrains.kotlin.psi.KtDestructuringDeclaration
+import org.jetbrains.kotlin.psi.KtDestructuringDeclarationEntry
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.psi.KtLambdaArgument
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtParameter
+import org.jetbrains.kotlin.psi.psiUtil.getChildOfType
+import org.jetbrains.kotlin.psi.psiUtil.getChildrenOfType
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.model.VarargValueArgument
+import org.jetbrains.kotlin.resolve.calls.smartcasts.getKotlinTypeForComparison
+import org.jetbrains.kotlin.resolve.calls.util.getResolvedCall
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.types.error.ErrorType
+
+
+enum class InlayKind(val base: InlayHintKind) {
+    TypeHint(InlayHintKind.Type),
+    ParameterHint(InlayHintKind.Parameter),
+    ChainingHint(InlayHintKind.Type),
+}
+
+private fun PsiElement.determineType(ctx: BindingContext): KotlinType? =
+    when (this) {
+        is KtCallExpression -> {
+            this.getKotlinTypeForComparison(ctx)
+        }
+        is KtParameter -> {
+            if (this.isLambdaParameter and (this.typeReference == null)) {
+                val descriptor = ctx[BindingContext.DECLARATION_TO_DESCRIPTOR, this] as CallableDescriptor
+                descriptor.returnType
+            } else null
+        }
+        is KtDestructuringDeclarationEntry -> {
+            val resolvedCall = ctx[BindingContext.COMPONENT_RESOLVED_CALL, this]
+            resolvedCall?.resultingDescriptor?.returnType
+        }
+        is KtProperty -> {
+            //TODO: better handling for unresolved-type error
+            val type = this.getKotlinTypeForComparison(ctx)
+            if (type is ErrorType) null else type
+        }
+        else -> null
+    }
+
+private fun PsiElement.hintBuilder(kind: InlayKind, file: CompiledFile, label: String? = null): InlayHint? {
+    val namedElement = ((this as? PsiNameIdentifierOwner)?.nameIdentifier ?: this)
+    val range = range(file.parse.text, namedElement.textRange)
+
+    val hint = when(kind) {
+        InlayKind.ParameterHint -> InlayHint(range.start, Either.forLeft("$label:"))
+        else ->
+            this.determineType(file.compile) ?.let {
+                InlayHint(range.end, Either.forLeft(
+                    "${(if (kind == InlayKind.TypeHint) ": " else "")}${DECL_RENDERER.renderType(it)}"
+                ))
+            } ?: return null
+    }
+    hint.kind = kind.base
+    hint.paddingRight = true
+    hint.paddingLeft = true
+    return hint
+}
+
+private fun callableArgsToHints(
+    callExpression: KtCallExpression,
+    file: CompiledFile,
+): List<InlayHint> {
+    val resolvedCall = callExpression.getResolvedCall(file.compile)
+
+    val hints = mutableListOf<InlayHint>()
+    resolvedCall?.valueArguments?.forEach { (t, u) ->
```

Comment:
You get a detekt (static code analysis warning) in your build for the nesting depth of the function here here. Maybe the inner label creation (line 92) could be extracted to its own helper function? 🙂